alacritty_terminal/src/keyboard_motion.rs

```diff
@@ -0,0 +1,767 @@
+    WordRightEnd,
+    /// Move to opposing bracket.
+    Bracket,
+}
+
+/// Cursor tracking keyboard motion position.
+#[derive(Default, Copy, Clone)]
+pub struct KeyboardCursor {
+    pub point: Point,
+}
+
+impl KeyboardCursor {
+    pub fn new(point: Point) -> Self {
+        Self { point }
+    }
+
+    /// Move keyboard motion cursor.
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub fn motion<T: EventListener>(mut self, term: &mut Term<T>, motion: KeyboardMotion) -> Self {
+        let display_offset = term.grid().display_offset();
+        let lines = term.grid().num_lines();
+        let cols = term.grid().num_cols();
+
+        let mut buffer_point = term.visible_to_buffer(self.point);
+
+        // Advance keyboard cursor
+        match motion {
+            KeyboardMotion::Up => buffer_point.line += 1,
+            KeyboardMotion::Down => buffer_point.line = buffer_point.line.saturating_sub(1),
+            KeyboardMotion::Left => {
+                buffer_point = expand_wide(term, buffer_point, true);
+                let wrap_point = Point::new(buffer_point.line + 1, cols - 1);
+                if buffer_point.col.0 == 0
+                    && buffer_point.line + 1 < term.grid().len()
+                    && is_wrap(term, wrap_point)
+                {
+                    buffer_point = wrap_point;
+                } else {
+                    buffer_point.col = Column(buffer_point.col.saturating_sub(1));
+                }
+            },
+            KeyboardMotion::Right => {
+                buffer_point = expand_wide(term, buffer_point, false);
+                if is_wrap(term, buffer_point) {
+                    buffer_point = Point::new(buffer_point.line - 1, Column(0));
+                } else {
+                    buffer_point.col = min(buffer_point.col + 1, cols - 1);
+                }
+            },
+            KeyboardMotion::First => {
+                buffer_point = expand_wide(term, buffer_point, true);
+                while buffer_point.col.0 == 0
+                    && buffer_point.line + 1 < term.grid().len()
+                    && is_wrap(term, Point::new(buffer_point.line + 1, cols - 1))
+                {
+                    buffer_point.line += 1;
+                }
+                buffer_point.col = Column(0);
+            },
+            KeyboardMotion::Last => buffer_point = last(term, buffer_point),
+            KeyboardMotion::FirstOccupied => buffer_point = first_occupied(term, buffer_point),
+            KeyboardMotion::High => {
+                let line = display_offset + lines.0 - 1;
+                buffer_point = Point::new(line, Column(0));
+            },
+            KeyboardMotion::Middle => {
+                let line = display_offset + lines.0 / 2;
+                buffer_point = Point::new(line, Column(0));
+            },
+            KeyboardMotion::Low => buffer_point = Point::new(display_offset, Column(0)),
```

Comment:
I've decided against this for the sake of simplicity.

Do you disagree with this decision? In theory this applies to nearly all bindings.

Since we do have a `first_occupied` function now, maybe it's worth making use of it?

Reply:
I'm fine with the current decision, but it'd be pretty simple to implement vim-like one if you return `Option<Point<usize>>` from `first_occupied` and do `unwrap_or(Point(line, Colunm(0))`.

Reply:
I've implemented the occupied motion for all modes that previously moved to `Column(0)`. It was pretty trivial to implement so I suppose it makes sense.